### DISCUSSION TITLE
get boot loops after changing bar size

### DISCUSSION THREAD

**Discussion by Black6spdZ** [terminatorul/NvStrapsReBar#13] [General]
I did everything according to instructions, everything compiled OK but if I set any size for BAR I get a boot-loop. One thing with my system is I don't have a PCI-to-PCI bridge on the same bus as the GPU, just a CPU root port?
Motherboard: Asus Z10PE-D8
CPUs: Xeon E5-2696v3
GPU: 2080 Super FTW3
![hwinfo](https://github.com/terminatorul/NvStrapsReBar/assets/6208317/9eadf0f7-2231-43c7-9557-1f738967c4b0)
[LocalPciGPU.h.txt](https://github.com/terminatorul/NvStrapsReBar/files/14228727/LocalPciGPU.h.txt)
[MONOLITH.txt](https://github.com/terminatorul/NvStrapsReBar/files/14228728/MONOLITH.txt)

**Comment by terminatorul** [comment]
Comparing with other screenshots, the root port looks normal. On my AMD motherboards it shows "PCIe GPP Bridge", don't know what it means. And your screenshot shows "Device Class: PCI-to-PCI Bridge" for the root port.

Your LocalPciGpu.h header file looks fine.

I can't tell what's wrong to cause the boot loop.

What values did you use in NvStrapsReBar ? Can you show a screenshot (before the first reboot) ? Can you try latest version from the bleed branch ?

**Reply by Black6spdZ** [reply]
I tried down to 1GB Bar size, but same boot loop until I clear cmos and reset the driver. Going to try the bleed branch now.

**Comment by Black6spdZ** [comment]
still getting the hang during post after first reboot "no video" with bleed
![nvstrapsrebar](https://github.com/terminatorul/NvStrapsReBar/assets/6208317/ce42837a-36b0-428c-ac0d-d0469a54dda8)

**Comment by terminatorul** [comment]
Can you make an USB stick with a Linux installer, for example Fedora or openSUSE ? You don't have to install it, only use the USB drive to start Linux, and then show the output of `lspci -vvv`, plus the hex content of the NvStrapsReBar variables that are listed with `ls /sys/firmware/efi/efivars/NvStraps*`

Or can you still boot Windows in safe mode ? That way you can just show the output from `NvStrapsReBar.exe`

**Reply by Black6spdZ** [reply]
boots and runs fine until a set a bar size.. I'm typing on it right now. booting to linux to get the info now. BTW thanks for the help!

**Comment by Black6spdZ** [comment]
got em 
[lspci_monolith.txt](https://github.com/terminatorul/NvStrapsReBar/files/14235076/lspci_monolith.txt)
[NvStrapsReBarStatus-e3ee4a27-e2a2-4435-bba3-184ccad935a8.dmp](https://github.com/terminatorul/NvStrapsReBar/files/14235077/NvStrapsReBarStatus-e3ee4a27-e2a2-4435-bba3-184ccad935a8.dmp)

**Comment by terminatorul** [comment]
There should be more then one EFI variable starting with NvStraps. Did you collect the .dmp after  you enabled ReBAR with `NvStrapsReBar.exe` (so when the boot loop triggered) ?

When you get the boot loop again, you can recover the system if you manually set back the current year in UEFI Setup to a value before 2023, and reboot. You can change back the year to the correct value after the reboot.

`lspci` on Linux shows ReBAR is not enabled, so the new DXE driver is not working.
The Status variable shows status 'Unconfigured', as if the `NvStrapsReBar` EFI variable is empty and you never enabled ReBAR...

After you enable ReBAR (with NvStrapsReBar.exe) and trigger the boot loop in Windows, you should see both `NvStrapsReBar` and `NvStrapsReBarStatus` variables under `/sys/firmware/efi/efivars/*`. Can you show them both please (collected at the same time)... ?

Also you should try the latest version from the github repository, it no longer uses hard-coded values in LocalPciGpu.h header.

**Comment by terminatorul** [comment]
Also, maybe your Xeon is for productivity or office use and you have the NVIDIA Studio drivers instead of GameReady drivers ? Can you check ?

**Reply by Black6spdZ** [reply]
when I said boot loop what I really meant is just a constant POST reset without video.. no OS load at all. Sorry, bad terminology. Normally video initializes right after POST code 98 on the boards diag display.. after enabling/sizing the BAR it reaches 98 then immediately jumps to 9C then 87 and hangs now.. if I power cycle it then just starts the POST loop. And no, no office or pro work with this machine so its using standard gameready driver.. was built 4 years ago to be the cheaper threadripper alternative

**Comment by terminatorul** [comment]
Oh, man, you have a dual-CPU server motherboard ... should have led with that. 

This hardware is ... intimidating to me. I do have an old Threadripper, but only because AMD marketed 1st gen Threadripper as a gaming computer (for lack of a better HEDT).

I think somehow I get the PCI bus or domain wrong, for such a system. Can you check what the boot CPU is ... and what PCI slot leads to what CPU ... ? Maybe just switch the GPU to a slot belonging to the other CPU ?

@xCuri0 can you help please: how do I make sure I am using the right PCI bus from the right CPU, on a dual-CPU server motherboard (Asus Z10PE-D8) ? Output from `lspci` is attached previously and appears to show a single PCI segment ... I suppose, I just don't know enough to be sure

@Black6spdZ  can you attach again the output from `lspci -t -vvv` ? With the `-t` for tree view this time, which might help

**Reply by terminatorul** [reply]
Can you check what slot you GPU is in ? Maybe you need to switch it from CPU2 to CPU1

![image](https://github.com/terminatorul/NvStrapsReBar/assets/378924/2ff2954f-0b34-4cb2-9967-d7699d24fb9c)

![image](https://github.com/terminatorul/NvStrapsReBar/assets/378924/8c691467-cd3a-4944-977f-3306d8463e4b)

**Reply by Black6spdZ** [reply]
it's in slot 1 for sure
[lspci-P.txt](https://github.com/terminatorul/NvStrapsReBar/files/14249582/lspci-P.txt)
[lspci-PP.txt](https://github.com/terminatorul/NvStrapsReBar/files/14249583/lspci-PP.txt)
[lspci-t.txt](https://github.com/terminatorul/NvStrapsReBar/files/14249584/lspci-t.txt)

system sidenote, I almost decided to go with the Ryzen 3950x but could see the writing on the wall, AM4 was long in the tooth and physically just too small to have the PCIe and MEM I/O  as the older HEDT Xeons at the time.. besides the turbo-boost unlock was just icing on the cake

**Comment by terminatorul** [comment]
PCI bus looks fine. 

This motherboard does not have ReBAR, and your PCI BAR size in the screenshot is set to 0 (System Default)

Did you set PCI BAR size for the GPU in `NvStrapsReBar.exe` (input value 64) ? And set Above 4G Decoding to Enabled and CSM to Disabled  in UEFI Setup ?
Show screenshot again with these settings please

**Comment by Black6spdZ** [comment]
yes, but once I set them system no longer POSTS or boots.. so I can't get anything meaningful.. should I try something very small? 256MB?

**Comment by xCuri0** [comment]
someone with the same board also can't use ReBarUEFI https://github.com/xCuri0/ReBarUEFI/issues/26


@terminatorul seems like there's nothing wrong with your module


kinda weird a server board has trouble with large BAR tho because usually they are paired with GPUs that require it like Nvidia Tesla / RTX A-series. maybe it's just a wrong bios setting

**Comment by Black6spdZ** [comment]
so, maybe new version is going somewhere.. I flashed it and it looks like everything was detected properly and set the recommended BAR size "8gb" in this case for my 2080S with 8gb and rebooted with fingers crossed.. it POSTS and booted fine! BUT it looks like it's not enabled? I do see a new file from the linux efi firmware folder though so maybe it's just a windows / driver issue?
[nvstraps.zip](https://github.com/terminatorul/NvStrapsReBar/files/14256527/nvstraps.zip)

**Reply by terminatorul** [reply]
Also add screenshot with the status code in NvStrapsReBar.exe

**Reply by Black6spdZ** [reply]
you got it
![status](https://github.com/terminatorul/NvStrapsReBar/assets/6208317/8605d077-1243-4e00-a7a4-e99d9a8eb860)

**Comment by Black6spdZ** [comment]
does this mean anything to you? some hidden options in bios hardcoded to certain values regarding MMIO
![mmio](https://github.com/terminatorul/NvStrapsReBar/assets/6208317/d2a33adf-2651-4784-a00f-c9b6ea64f8c8)

**Comment by Black6spdZ** [comment]
look what I found hidden away
![bar-stuff](https://github.com/terminatorul/NvStrapsReBar/assets/6208317/95398349-95b0-47c6-88da-9e19cda306b2)

**Reply by terminatorul** [reply]
The status code from the driver is good. But the motherboard does not enable ReBAR out of the box, even if GPU advertises ReBAR support on the PCI bus.

Can you find the right port in your screenshot here for the GPU in slot 1 (CPU 1) ?

Do you see the same settings when you enter UEFI setup ? Looks like they allow you to manually set the BAR size for each PCI device ! But it might be for integrated devices only (from the chipset)

**Comment by Black6spdZ** [comment]
so I had to move my GPU to get it into that Port3A "PCIe Slot 3 on mobo" because only that port and socket 2 port 3A "slot 7" have those BAR size adjustments.. who would have known, not documented anywhere and had to spend hours digging through bios to find and unhide it. I haven't made any bios adjustments yet but just swapping slots I still get same problem, system boots but no bar size change.. nvstrapsrebar shows: UEFI DXE driver status: GPU-side ReBAR Configured with PCI confirm (0x0300000000000064)
Target PCI BAR size: 0 / System default. any idea what I should change in the bios with regards to those BAR sizes?

**Reply by terminatorul** [reply]
I was hoping you can find a BAR1 size or BAR12 size option... 
Can you just disable split BARs, maybe it will make the MB stick to the 64-bit BAR

What are NTB BARs, can you show the Control Help String at the buttom ?

**Reply by xCuri0** [reply]
@Black6spdZ sounds like you're still using the old bridge config which won't work in the new slot

**Reply by Black6spdZ** [reply]
yeah, that gave me a hang until I cleared CMOS.. was strange that it initially booted to windows fine.. I did what I thought was clearing the old config and resetting the bar size again, rebooted and it hung. gave it a CMOS clear and it was fine but back to it's old ways, BAR is stuck at 256MB

**Reply by Black6spdZ** [reply]
NTB is a mechanism for direct PCIe networking between servers! https://www.dolphinics.com/products/pcie_Intel_NTB_networking.html
https://www.dolphinics.com/products/PXH810.html

**Comment by Black6spdZ** [comment]
Non-Transparent Bridge. I've tried everything there: disable NTB bridge bars, enable/disable split bars, changed the sizes.. no change. Still in the same boat :( maybe it just wasn't meant to be. I'm not sure what else I can do

**Reply by terminatorul** [reply]
Sorry, I do not know what else to do either. Looks like the firmware does not like the trick NvStrapsReBar (and ReBarUEFI) are doing to force ReBAR on a non-supported motherboard (they select and write the target BAR size in advance, before the stock firmware, that normally enumerates the PCI, can select it ... and happily the stock firmware is usually smart enough to not override the pre-existing value, and go with it and allocate the new size)

Maybe you can make a support request to Asus ... I believe the board is old now, so not sure that works. Make sure you start with something like "dual-CPU server motherboard", add the screenshot, and say you have "a device" that needs 8GB BAR1 size ...

**Reply by Black6spdZ** [reply]
what about a UEFI based driver? would that load after system initialization?

**Reply by terminatorul** [reply]
Amazingly, I don't know what a UEFI based driver is (other then the DXE driver in NvStrapsReBar, but it looks like you want something different)

**Reply by terminatorul** [reply]
Also maybe there is a support forum for this, or other forum like win-raid...

**Reply by Black6spdZ** [reply]
yep, an efi file is included with your compile. going to do this:

Use EFI Shellx64 or other suitable UEFI shell to set automatic load (and execution) of NvStrapsReBar.efi during system boot. Use cmd ' bcfg driver add 0 fs1:\EFI\Boot\NvStrapsReBar.efi "NvStrapsReBar" ' where 'fs1:\EFI\Boot\NvStrapsReBar.efi' is path to DXE driver file on UEFI boot partition

**Reply by Black6spdZ** [reply]
does it matter where it gets placed in the bios?

**Comment by Black6spdZ** [comment]
removed the bios ffs driver, then used the UEFI shell to copy over and add efi the driver.. it appears it worked.. at least no freezes or bsod's but still stuck with 256MB BAR but did get a different diag code in nvstrapsrebar.exe

UEFI DXE driver status: Configured (0x0000000000000014)
Target PCI BAR size: 0 / System default
![20240213_111009](https://github.com/terminatorul/NvStrapsReBar/assets/6208317/81d94fea-2685-4a2a-b82a-eb3a54ab43f9)

if this method works it would be a safer way for others to load the patch without flashing bios

**Reply by xCuri0** [reply]
@Black6spdZ that method of bcfg does not work with NvStrapsRebar or ReBarUEFI

Because the module has to be loaded before PCI enumeration. In this case it's running after because the firmware needs all that stuff initialized in order for it to load a module from storage.

**Comment by Black6spdZ** [comment]
oh.. well damn, wishful thinking on my part

**Comment by Black6spdZ** [comment]
would doing a boot with my asus q-code logger offer any useful information?

**Comment by Black6spdZ** [comment]
I'm confused with HWInfo's output here 
![gpu-info](https://github.com/terminatorul/NvStrapsReBar/assets/6208317/6925b5be-1fea-44cb-9100-5d20e8e48d13)

**Comment by Black6spdZ** [comment]
would this switch cause issues??
![block diagram](https://github.com/terminatorul/NvStrapsReBar/assets/6208317/6259edeb-041e-48b8-b5ed-b5d844898fee)

**Comment by ghost** [comment]
That affects PCIe lanes, not BAR1 size. Your device manager pic before shows BAR1 in 32-bit space, with above 4G decoding it should be above 4G unless a 4G display patch is in place. What is Load RT32 Image ? Would you try disabling it in BIOS setup and see if BAR1 address moves. Do it without setting Rebar first, ie 256MB BAR's.

**Comment by Black6spdZ** [comment]
> That affects PCIe lanes, not BAR1 size. Your device manager pic before shows BAR1 in 32-bit space, with above 4G decoding it should be above 4G unless a 4G display patch is in place. What is Load RT32 Image ? Would you try disabling it in BIOS setup and see if BAR1 address moves. Do it without setting Rebar first, ie 256MB BAR's.

disabled "Load RT32 Image".. I'm not seeing any change at all. what do you mean by 4G display patch?

![rt32_disabled](https://github.com/terminatorul/NvStrapsReBar/assets/6208317/94a8afe5-9a4f-4c88-8688-bc9f7d69b2cc)

**Comment by ghost** [comment]
I guess some display adapters / OS don't play well above 4G so an option is available to use them with 64-bit addressing but below 4GiB.

Example on my X99 with 2696 v3
![BIOS4G](https://github.com/terminatorul/NvStrapsReBar/assets/147298897/718ebd4f-0235-4068-892a-4c0271fa53f1)

With 4G and patch enabled
![4GPatch](https://github.com/terminatorul/NvStrapsReBar/assets/147298897/e64d99c1-e119-442a-ba76-2355b912f619)

And with patched disabled addressing for 64-bit BAR1 and BAR2 are moved to 1TiB address range while 32-bit BAR0 remains below 4GiB
![4GNoPatch](https://github.com/terminatorul/NvStrapsReBar/assets/147298897/6144b5dd-7fb0-4ef5-836e-b1b4dc884bd9)

Your device manager pic is not clear as it appears to show multiple entries. There can be old records but usually these are a lighter shade but yours look the same? Maybe need to DDU (Display Driver Uninstaller) and start again without Rebar and get above 4G working with your single RTX card?

**Comment by Black6spdZ** [comment]
I don't have that "above 4G decoding patch" option.. searched bios with UEFITool
this may be the key.. is it something included the the UEFIPatch tool "patches.txt" that possibly is slightly different because of my dual board? I'm digging on it now for info
![mem range - bar sizes](https://github.com/terminatorul/NvStrapsReBar/assets/6208317/7fd56446-91c1-450c-8d94-2ccea085e8b3)

**Comment by Black6spdZ** [comment]
from https://winraid.level1techs.com/t/help-needed-asrock-x99-taichi-modding-rebar-cpu-microcodes-default-bios-values/91869/7

rebar 4Gb => boot, but no bios display until windows
rebar 2Gb => NO boot at all, bios is bootlooping. 
rebar Unlimited => boot, but no bios display until windows
last try : “above 4G decoding patch OFF” and rebar unlimited => boot, and I have bios display working !

seems I am late to this party https://github.com/xCuri0/ReBarUEFI/issues/4#issuecomment-1364450119

**Comment by Black6spdZ** [comment]
eyelids are getting heavy.. if anyone wants to doublecheck my searching.. maybe I've overlooked something. I can't even find what that "patch" does, but it plagues many boards that don't have it as well from the searches
[Z10PEWS.zip](https://github.com/terminatorul/NvStrapsReBar/files/14321030/Z10PEWS.zip)
